### PR TITLE
Adds SSE support

### PR DESCRIPTION
**Pull request by smouli**
https://github.com/minio/minio-py/issues/682

### CODE REVIEW

File: minio/sse.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            "X-Amz-Server-Side-Encryption-Customer-Key": b64key.decode(),
+            "X-Amz-Server-Side-Encryption-Customer-Key-MD5": md5_key
+            }
+
+        return keys
+
+class copy_SSEC(object):
+
+    def __init__(self, key):
+        self.key = key
```
Comment (poornas) [inline minio/sse.py]: if key is not 32 bytes, need to throw error - SSE-C keys need to be 256 bit base64 encoded.

File: tests/functional/tests.py [hunk 2]
```diff
@@ -1702,6 +1707,9 @@ def main():
             with open(largefile, 'wb') as file_data:
                 shutil.copyfileobj(LimitedRandomReader(11*1024*1024), file_data)
 
+        cust_key = b'AABBCCDDAABBCCDDAABBCCDDAABBCCDD'
+        ssec = SSEC(cust_key)
+
```
Comment (poornas) [inline tests/functional/tests.py]: add a test with invalid shorter key as well.

File: tests/functional_test.sh [hunk 3]
```diff
@@ -60,5 +60,5 @@ main () {
 
 # invoke the script
 # Move to the directory which contains this script and invoke it
-cd $(dirname $(realpath $0)) ; \
+#cd $(dirname $(realpath $0)) ; \
```
Comment (poornas) [inline tests/functional_test.sh]: @smouli, this needs to be uncommented

File: examples/put_and_get_object_sse-c.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
+# -*- coding: utf-8 -*-
+# Minio Python Library for Amazon S3 Compatible Cloud Storage, (C) 2015 Minio, Inc.
```
Comment (kannappanr) [inline examples/put_and_get_object_sse-c.py]: Change the year from `2015` to `2018`

File: examples/put_object_sse-kms.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
+# -*- coding: utf-8 -*-
+# Minio Python Library for Amazon S3 Compatible Cloud Storage, (C) 2015 Minio, Inc.
```
Comment (kannappanr) [inline examples/put_object_sse-kms.py]: change the year from `2015` to `2018`

File: examples/put_object_sse-s3.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
+# -*- coding: utf-8 -*-
```
Comment (kannappanr) [inline examples/put_object_sse-s3.py]: change the year from 2015 to 2018

File: minio/api.py [hunk 7]
```diff
@@ -612,7 +614,7 @@ def fget_object(self, bucket_name, object_name, file_path, request_headers=None)
         # Return the stat
         return stat
 
-    def get_object(self, bucket_name, object_name, request_headers=None):
+    def get_object(self, bucket_name, object_name, request_headers=None, sse=None):
         """
```
Comment (kannappanr) [inline minio/api.py]: Here you should check if the see parameter passed is `SSE-C`, if not you should error out I guess

File: minio/api.py [hunk 8]
```diff
@@ -612,7 +614,7 @@ def fget_object(self, bucket_name, object_name, file_path, request_headers=None)
         # Return the stat
         return stat
 
-    def get_object(self, bucket_name, object_name, request_headers=None):
+    def get_object(self, bucket_name, object_name, request_headers=None, sse=None):
```
Comment (kannappanr) [inline minio/api.py]: Here you should check if the see parameter passed is SSE-C, if not you should error out I guess

File: minio/api.py [hunk 9]
```diff
@@ -636,9 +638,10 @@ def get_object(self, bucket_name, object_name, request_headers=None):
 
         return self._get_partial_object(bucket_name,
                                         object_name,
-                                        request_headers=request_headers)
+                                        request_headers=request_headers,
+                                        sse=sse)
 
-    def get_partial_object(self, bucket_name, object_name, offset=0, length=0, request_headers=None):
+    def get_partial_object(self, bucket_name, object_name, offset=0, length=0, request_headers=None, sse=None):
```
Comment (kannappanr) [inline minio/api.py]: Same comment as above

File: minio/api.py [hunk 10]
```diff
@@ -701,10 +717,12 @@ def copy_object(self, bucket_name, object_name, object_source,
...
 
         return parse_copy_object(bucket_name, object_name, response.data)
-
+    
     def put_object(self, bucket_name, object_name, data, length,
                    content_type='application/octet-stream',
-                   metadata=None):
+                   metadata=None, 
+                   sse=None,
+                   ):
```
Comment (kannappanr) [inline minio/api.py]: Add another check here like

```
if sse:  #sse argument to put_object has to be of these three types
            if sse.type() == "SSE-C" or sse.type() == "SSE-KMS" or sse.type() == "SSE-S3":
                headers.update(sse.marshal())
            else:
                raise InvalidArgumentError("unsuported type of sse argument in put_object")
```

File: minio/api.py [hunk 11]
```diff
@@ -894,19 +913,25 @@ def list_objects_v2(self, bucket_name, prefix=None, recursive=False):
...
         Check if an object exists.
 
         :param bucket_name: Bucket of object.
         :param object_name: Name of object
         :return: Object metadata if object exists
         """
+
+        headers = {}
+
+        if sse:
```
Comment (kannappanr) [inline minio/api.py]: The only valid sse type here is `SSE-C`, you should check if it is of that type, if it is passed

File: minio/api.py [hunk 12]
```diff
@@ -665,16 +671,18 @@ def get_partial_object(self, bucket_name, object_name, offset=0, length=0, reque
         :return: :class:`urllib3.response.HTTPResponse` object.
 
         """
+        #is_valid_sse_object(sse=sse)
```
Comment (kannappanr) [inline minio/api.py]: This commented line needs to be removed

File: minio/api.py [hunk 13]
```diff
@@ -695,6 +703,14 @@ def copy_object(self, bucket_name, object_name, object_source,
         headers = {}
         if conditions:
             headers = {k: v for k, v in conditions.items()}
+        
+        # Source argument to copy_object can only be of type copy_SSE_C
+        is_valid_sse_object(source_sse)
+        headers.update(source_sse.marshal())
+        
+        #Destination argument to copy_object cannot be of type copy_SSE_C
+        is_valid_sse_put_object(sse)
```
Comment (kannappanr) [inline minio/api.py]: This name can be `is-valid_sse_object`

File: minio/api.py [hunk 14]
```diff
@@ -695,6 +703,14 @@ def copy_object(self, bucket_name, object_name, object_source,
         headers = {}
         if conditions:
             headers = {k: v for k, v in conditions.items()}
+        
+        # Source argument to copy_object can only be of type copy_SSE_C
+        is_valid_sse_object(source_sse)
```
Comment (kannappanr) [inline minio/api.py]: this can be named `is_valid_source_sse_object`

File: minio/api.py [hunk 15]
```diff
@@ -731,8 +749,14 @@ def put_object(self, bucket_name, object_name, data, length,
             with your PUT request.
         :return: etag
         """
+        headers = {}
+        if sse:
+            is_valid_sse_put_object(sse)
```
Comment (kannappanr) [inline minio/api.py]: this can be named `is_valid_sse_put_object`

File: minio/api.py [hunk 16]
```diff
@@ -896,19 +921,25 @@ def list_objects_v2(self, bucket_name, prefix=None, recursive=False):
...
-    def stat_object(self, bucket_name, object_name):
+    def stat_object(self, bucket_name, object_name, sse=None):
         """
         Check if an object exists.
 
         :param bucket_name: Bucket of object.
         :param object_name: Name of object
         :return: Object metadata if object exists
         """
+        is_valid_sse_object(sse=sse)
```
Comment (kannappanr) [inline minio/api.py]: this needs to be moved under the `if sse:` part of code

File: minio/api.py [hunk 17]
```diff
@@ -1364,6 +1397,7 @@ def _get_partial_object(self, bucket_name, object_name,
         :return: :class:`urllib3.response.HTTPResponse` object.
 
         """
+        is_valid_sse_object(sse=sse)
```
Comment (kannappanr) [inline minio/api.py]: This needs to be named `is_valid_sse_customer_object`

File: tests/functional/tests.py [hunk 18]
```diff
@@ -672,10 +675,11 @@ def validate_stat_data(st_obj, expected_size, expected_meta):
         raise ValueError('Incorrect file type. Expected: is_dir=False',
                          ', received: is_dir=', received_is_dir)
 
-    if not all(i in expected_meta.items() for i in received_metadata.items()):
+    #if not all(i in expected_meta.items() for i in received_metadata.items()):
```
Comment (poornas) [inline tests/functional/tests.py]: this commented line needs to be removed

Comment (smouli) [reply tests/functional/tests.py]: i removed it

**Pull request merged**